connectors/sources/slack.py

```diff
@@ -0,0 +1,429 @@
+        elif user_profile.get("real_name_normalized"):
+            return user_profile["real_name_normalized"]
+        elif user.get("real_name"):
+            return user["real_name"]
+        elif user.get("name"):
+            return user["name"]
+        else:
+            return user["id"]
+
+    def adapt_user(self, user):
+        user_profile = user.get("profile", {})
+        user["real_name_normalized"] = user_profile.get("real_name_normalized")
+        user["profile_real_name"] = user_profile.get("real_name")
+        user["display_name"] = user_profile.get("display_name")
+        user["display_name_normalized"] = user_profile.get("display_name_normalized")
+        user["first_name"] = user_profile.get("first_name")
+        user["last_name"] = user_profile.get("last_name")
+        user = self._dict_slice(
+            user,
+            (
+                "id",
+                "name",
+                "deleted",
+                "real_name",
+                "real_name_normalized",
+                "profile_real_name",
+                "display_name",
+                "display_name_normalized",
+                "first_name",
+                "last_name",
+                "is_admin",
+                "is_owner",
+                "is_bot",
+            ),
+        )
+        user["_id"] = user["id"]
+        user["type"] = "user"
+        return user
+
+    def adapt_channel(self, channel, usernames):
+        topic = channel.get("topic", {})
+        topic_creator_id = topic.get("creator")
+        purpose = channel.get("purpose", {})
+        purpose_creator_id = purpose.get("creator")
+        channel["topic"] = topic.get("value")
+        channel["topic_author"] = usernames.get(topic_creator_id, topic_creator_id)
+        channel["topic_last_updated"] = topic.get("last_set")
+        channel["purpose"] = purpose.get("value")
+        channel["purpose_author"] = usernames.get(
+            purpose_creator_id, purpose_creator_id
+        )
+        channel["purpose_last_updated"] = purpose.get("last_set")
+        channel["last_updated"] = channel.get("updated")
+        channel = self._dict_slice(
+            channel,
+            (
+                "id",
+                "name",
+                "name_normalized",
+                "topic",
+                "topic_author",
+                "topic_last_updated",
+                "purpose",
+                "purpose_author",
+                "purpose_last_updated",
+                "last_updated",
+                "created",
+            ),
+        )
+        channel["_id"] = channel["id"]
+        channel["type"] = "channel"
+        return channel
+
+    def adapt_message(self, message, channel, usernames):
+        user_id = message.get("user", message.get("bot_id"))
+        message["author"] = usernames.get(user_id, user_id)
+        message["text"] = re.sub(
+            r"<@([A-Z0-9]+)>", self.convert_usernames, message["text"]  # type: ignore
+        )
```

Comment:
Idk how it works in Python 100%, but I think we might wanna compile this regex to speed things up. But might not be a big deal too?

Reply:
IDK how it works in python either. I stole this approach from SO, but I can definitely put in more research. Then maybe I can remove the `# type: ignore` (pyright was unhappy)